src/pint/observatory/__init__.py

```diff
@@ -549,3 +559,131 @@ def convert_angle(x):
             # Check ITOA code?
             # Check time corrections?
     return report
+
+
+def check_for_new_clock_files_in_tempo12_repos(update_download=True, show_diff=100):
+    """Try to determine whether PINT's clock files are up to date.
+
+    This iterates through all observatories for which PINT has clock corrections
+    and checks the clock corrections PINT is using against those in the
+    TEMPO/TEMPO2 repositories. Ones that differ are reported, along with
+    a little context.
+
+    The web versions are actually downloaded into the astropy cache, so
+    optional automated updates would be possible.
+
+    This function is still a prototype.
+
+    Parameters
+    ----------
+    update_download : bool
+        If True, download new versions even if there are versions in the
+        astropy cache. (they might be old.)
+    show_diff : int
+        Show the difference between local and remote files if it is less than
+        this many lines. (Huge diffs are probably just more correction data.)
+    """
+    import collections
+    import difflib
+    import astropy.utils.data
+
+    # Ensure all observatories are loaded and warnings are emitted
+    clock_files = collections.defaultdict(list)
+    for a, o in Observatory._registry.items():
+        if not hasattr(o, "clock_file"):
+            log.debug(f"Skipping bogus clock file for observatory {a}")
+            continue
+        try:
+            # Ensure clock corrections are loaded and suppress some warnings
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                o.clock_corrections(Time(99999, format="mjd"), limits="error")
+        except RuntimeError:
+            pass
+        if o.clock_file is None or o.clock_file == "time.dat":
+            log.debug(f"Skipping bogus clock file for observatory {a}: {o.clock_file}")
+            continue
+        if isinstance(o.clock_fullpath, str):
+            cfs = [o.clock_fullpath]
+        else:
+            cfs = o.clock_fullpath
+        for clock_file in cfs:
+            if not os.path.isfile(clock_file):
+                log.debug(
+                    f"Skipping unreadable clock file for observatory {a}: {clock_file}"
+                )
+                continue
+            clock_files[clock_file].append(o)
+
+    tempo_repo = "https://raw.githubusercontent.com/nanograv/tempo/master/clock/"
```

Comment:
What is up with that tempo repo?  The official TEMPO is on Sourceforge.  The clock dir is:  https://sourceforge.net/p/tempo/tempo/ci/master/tree/clock/  Although pulling without a "\?format\=raw" on the end gives you a bunch of html crap.

Reply:
Ah, I thought github was the official one. Didn't Sourceforge start including malware in built binaries?